File: frontend/src/types/compatibility.ts
// Shared compatibility types for frontend and backend
// Converted from backend/app/schemas/compatibility.py

export enum MatchCategory {
  EXCELLENT = "excellent",  // > 80%
  GOOD = "good",           // 60-80%
  FAIR = "fair",           // 40-60%
  POOR = "poor"            // < 40%
}

export interface SkillMatch {
  name: string;
  score: number; // 0-1
  category: string;
  semantic_score?: number;
  keyword_score?: number;
  is_matched: boolean;
  suggestions?: string[];
}

export interface RoleMatch {
  title_similarity: number; // 0-1
  category_match: number; // 0-1
  required_level?: string;
  current_level?: string;
  overall_score: number; // 0-1
  category: MatchCategory;
}

export interface ExperienceMatch {
  years_match: number; // 0-1
  role_relevance: number; // 0-1
  required_years: number;
  actual_years: number;
  overall_score: number; // 0-1
  category: MatchCategory;
  gaps?: string[];
}

export interface CompatibilityReport {
  resume_id: string; // UUID as string
  job_id: string; // UUID as string
  overall_score: number; // 0-1
  category: MatchCategory;
  
  // Detailed matches
  skill_matches: SkillMatch[];
  role_match: RoleMatch;
  experience_match: ExperienceMatch;
  
  // Summary statistics
  matched_skills_count: number;
  missing_skills_count: number;
  average_skill_score: number;
  
  // Improvement suggestions
  suggestions: string[];
  
  // Additional metadata
  analysis_timestamp: string;
  match_strategy: string; // "hybrid", "semantic", or "keyword"
}

export interface CompatibilityRequest {
  resume_id: string; // UUID as string
  job_id: string; // UUID as string
  strategy: string; // "hybrid", "semantic", or "keyword"
}

export interface CompatibilityResponse {
  report: CompatibilityReport;
  processing_time: number; // in seconds
  cached: boolean;
} 

File: backend/app/schemas/compatibility.py
from pydantic import BaseModel, Field
from typing import List, Dict, Optional
from uuid import UUID
from enum import Enum

class MatchCategory(str, Enum):
    EXCELLENT = "excellent"  # > 80%
    GOOD = "good"          # 60-80%
    FAIR = "fair"          # 40-60%
    POOR = "poor"          # < 40%

class SkillMatch(BaseModel):
    name: str
    score: float = Field(ge=0, le=1)
    category: str
    semantic_score: Optional[float] = None
    keyword_score: Optional[float] = None
    is_matched: bool
    suggestions: Optional[List[str]] = None

class RoleMatch(BaseModel):
    title_similarity: float = Field(ge=0, le=1)
    category_match: float = Field(ge=0, le=1)
    required_level: Optional[str] = None
    current_level: Optional[str] = None
    overall_score: float = Field(ge=0, le=1)
    category: MatchCategory

class ExperienceMatch(BaseModel):
    years_match: float = Field(ge=0, le=1)
    role_relevance: float = Field(ge=0, le=1)
    required_years: float
    actual_years: float
    overall_score: float = Field(ge=0, le=1)
    category: MatchCategory
    gaps: Optional[List[str]] = None

class CompatibilityReport(BaseModel):
    resume_id: UUID
    job_id: UUID
    overall_score: float = Field(ge=0, le=1)
    category: MatchCategory
    
    # Detailed matches
    skill_matches: List[SkillMatch]
    role_match: RoleMatch
    experience_match: ExperienceMatch
    
    # Summary statistics
    matched_skills_count: int
    missing_skills_count: int
    average_skill_score: float
    
    # Improvement suggestions
    suggestions: List[str]
    
    # Additional metadata
    analysis_timestamp: str
    match_strategy: str = "hybrid"  # hybrid, semantic, or keyword

class CompatibilityRequest(BaseModel):
    resume_id: UUID
    job_id: UUID
    strategy: str = "hybrid"

class CompatibilityResponse(BaseModel):
    report: CompatibilityReport
    processing_time: float  # in seconds
    cached: bool = False 